### DISCUSSION TITLE
What's the best way to hide block text attributes in modelspace ?

### DISCUSSION THREAD

**Discussion by Gorgious56** [mozman/ezdxf#648] [Q&A] [answered]
I'm writing a script where I initialize blocks and text attributes inside them but I don't want the text attribs to appear in modelspace. I've found that the [invisible](https://ezdxf.readthedocs.io/en/stable/dxfentities/dxfgfx.html#ezdxf.entities.DXFGraphic.dxf.invisible) attribute seems to do what I'm going after, but I can't make it work. I'm opening the file in Acad afterwards and the text is still visible.

Here's a simple script that creates a blockdef, instance a blockref and populate its attribute. I expected the text not to appear in modelspace. I've looked around a bit but couldn't find a workaround.

```
import ezdxf

dxf_file = ezdxf.new(dxfversion="R2010")
block_def = dxf_file.blocks.new(name="Test")
block_def.add_attdef("test_attr", dxfattribs={"invisible": 1})
block_instance = dxf_file.modelspace().add_blockref("Test", (0, 0, 0))
block_instance.add_attrib("test_attr", "test")

dxf_file.saveas(r"C:\path\to\file.dxf")
```

Thanks for the awesome library btw ! It saves me countless hours of tedious work :)

**Comment by mozman** [comment]
By using `add_attrib()` you create a new ATTRIB from scratch, so you have to set **all** DXF attributes by yourself including the  location, in your example, you also have to set the `invisible` flag like you did for the ATTDEF entity.

But I assume this is what you want:

```Python
    block_instance.add_auto_attrib({"test_attr", "test"})
```

This adds new ATTRIB entities based on template ATTDEF entities at the same relative location as the ATTDEF and the same settings.

See [docs](https://ezdxf.mozman.at/docs/blocks/insert.html#ezdxf.entities.Insert.add_auto_attribs) for more information.

**Reply by Gorgious56** [reply]
Hello ! Thanks for the quick reply. :)

I have to say I have tested pretty much all matters of combinations, including `add_auto_attribs` but none of them seem to be setting the invisible flag to the attrib when I open the file.

If I follow correctly this should create a block with no geometry and an invisible attribute in modelspace, but I can still see the attribute :

```
import ezdxf

dxf_file = ezdxf.new(dxfversion="R2010")
block_def = dxf_file.blocks.new(name="Test")
block_instance = dxf_file.modelspace().add_blockref("Test", (0, 0, 0))
block_instance.add_attrib("test_attr", "test", dxfattribs={"invisible": 1})
dxf_file.saveas(r"C:\path\to\file.dxf")
```

Am I still missing something ?

**Comment by mozman** [comment, marked as answer]
The `invisible` flag is the usual flag to set DXF entities invisible and seems not to work for ATTRIB entities.
The ATTRIB entity has a special visibility flag which is stored in the DXF `flag`  attribute but can be set by the property `is_invisible`:

```Python
...

block_instance = dxf_file.modelspace().add_blockref("Test", (0, 0, 0))
attrib = block_instance.add_attrib("test_attr", "test")
attrib.is_invisible = True
dxf_file.saveas(r"C:\Users\manfred\Desktop\Outbox\attribs.dxf")
```

**Comment by Gorgious56** [comment]
Awesome, that did the trick ! Thanks :)

For reference this is where I'm using this : https://github.com/Gorgious56/dotbim_io_dxf
It's a converter script between [dotbim](https://github.com/paireks/dotbim) and dxf.

Cheers !

**Reply by mozman** [reply]
dotbim is a nice idea and the complexity of IFC is really overwhelming, but BIM is much more than exchange of 3D meshes ...

**Reply by Gorgious56** [reply]
You're right, the naming is arguably a bit unfortunate but the project seems to receive a decent amount of traction and I see it as a good stepping stone in the world of bim & collaboration. We'll see :)
